apiserver/common/secrets/secrets.go

```diff
@@ -580,71 +584,77 @@ func removeSecrets(
 	removeState SecretsRemoveState, adminConfigGetter BackendAdminConfigGetter,
 	authTag names.Tag, args params.DeleteSecretArgs,
 	canDelete func(*coresecrets.URI) error,
-	removeSecretsFromBackend func(provider.SecretBackendProvider, provider.ModelBackendConfig, []string) error,
+	removeFromBackend func(provider.SecretBackendProvider, provider.ModelBackendConfig, provider.SecretRevisions) error,
 ) (params.ErrorResults, error) {
-	removeSecret := func(uri *coresecrets.URI, revisions ...int) ([]coresecrets.ValueRef, error) {
-		if _, err := removeState.GetSecret(uri); err != nil {
-			// Check if the uri exists or not.
-			return nil, errors.Trace(err)
-		}
-		if err := canDelete(uri); err != nil {
-			return nil, errors.Trace(err)
-		}
-		return removeState.DeleteSecret(uri, revisions...)
-	}
-
-	type deleteInfo struct {
-		uri       *coresecrets.URI
-		revisions []int
-	}
-	toDelete := make([]deleteInfo, len(args.Args))
-	for i, arg := range args.Args {
-		uri, err := coresecrets.ParseURI(arg.URI)
-		if err != nil {
-			return params.ErrorResults{}, errors.Trace(err)
-		}
-		toDelete[i] = deleteInfo{uri: uri, revisions: arg.Revisions}
-	}
 	result := params.ErrorResults{
 		Results: make([]params.ErrorResult, len(args.Args)),
 	}
-	externalRevisions := make(map[string]provider.SecretRevisions)
-	for i, d := range toDelete {
-		external, err := removeSecret(d.uri, d.revisions...)
```

Comment:
The above `external` returned from removeSecret func is always empty for unit/app-owned secrets which means that the code below for removing secrets from the backend is actually a no-op. 
The uniter removes secret resources from the backend before removing them from the state. To keep consistency, we now do the same way for user secret revision removal.